iothub/service/tests/DigitalTwinsClientTests.cs

```diff
@@ -0,0 +1,354 @@
+        {
+            Protocol = IotHubTransportProtocol.Tcp,
+            RetryPolicy = new IotHubServiceNoRetry()
+        };
+
+        [TestMethod]
+        public async Task DigitalTwinsClient_GetAsync()
+        {
+            // arrange
+            string digitalTwinId = "foo";
+            var digitalTwin = new BasicDigitalTwin
+            {
+                Id = digitalTwinId,
+            };
+
+            var mockCredentialProvider = new Mock<IotHubConnectionProperties>();
+            mockCredentialProvider
+                .Setup(getCredential => getCredential.GetAuthorizationHeader())
+                .Returns(s_validMockAuthenticationHeaderValue);
+            var mockHttpRequestFactory = new HttpRequestMessageFactory(s_httpUri, "");
+
+            using var mockHttpResponse = new HttpResponseMessage
+            {
+                StatusCode = HttpStatusCode.OK,
+                Content = HttpMessageHelper.SerializePayload(digitalTwin),
+            };
+            mockHttpResponse.Headers.Add("ETag", "\"AAAAAAAAAAE=\"");
+            
+            var mockHttpClient = new Mock<HttpClient>();
+            mockHttpClient
+                .Setup(restOp => restOp.SendAsync(It.IsAny<HttpRequestMessage>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(mockHttpResponse);
+
+            var digitalTwinsClient = new DigitalTwinsClient(
+                HostName,
+                mockCredentialProvider.Object,
+                mockHttpClient.Object,
+                mockHttpRequestFactory,
+                s_retryHandler);
+
+            // act
+            Func<Task> act = async () => await digitalTwinsClient.GetAsync<BasicDigitalTwin>(digitalTwinId).ConfigureAwait(false);
+
+            // assert
+            await act.Should().NotThrowAsync();
+        }
+
+        [TestMethod]
+        public async Task DigitalTwinsClient_GetAsync_NullTwinIdThrows()
+        {
+            // arrange
+            string digitalTwinId = "";
+            using var serviceClient = new IotHubServiceClient(s_connectionString);
+            DigitalTwinsClient digitalTwinsClient = serviceClient.DigitalTwins;
+
+            // act
+            Func<Task> act = async () => await digitalTwinsClient.GetAsync<BasicDigitalTwin>(digitalTwinId);
+
+            // assert
+            await act.Should().ThrowAsync<ArgumentException>();
+        }
+
+        [TestMethod]
+        public async Task DigitalTwinsClient_GetAsync_HttpException()
+        {
+            // arrange
+            string digitalTwinId = Guid.NewGuid().ToString();
+            using var serviceClient = new IotHubServiceClient(s_connectionString);
+            DigitalTwinsClient digitalTwinsClient = serviceClient.DigitalTwins;
+
+            // act
+            // deliberately throw http exception by searching for twin that does not exist
+            Func<Task> act = async () => await digitalTwinsClient.GetAsync<string>(digitalTwinId);
```

Comment:
We haven't mocked the connection, so I suspect this exception is because the creds are made up. No?

Reply:
As a unit test, this should fail because you've mocked it to fail, not because you've used a GUID for the digital twin Id that we assume wouldn't exist.

Reply:
Same with other tests in this class.

I recommend using Fiddler to ensure no outbound HTTP connections are attempted.